packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+        : package{package}, insideTestFile{ctx.file.data(ctx).isPackagedTest()} {}
+
+    // `keep-def` will reference constants in a way that looks like a packaging violation, but is actually fine. This
+    // boolean allows for an early exit when we know we're in the context of processing one of these sends. Currently
+    // the only sends that we process this way will not have any nested method calls, but if that changes this will need
+    // to become a stack.
+    bool ignoreConstant = false;
+
+    ast::ExpressionPtr preTransformSend(core::Context ctx, ast::ExpressionPtr tree) {
+        auto &send = ast::cast_tree_nonnull<ast::Send>(tree);
+        this->ignoreConstant = send.fun == core::Names::keepForIde();
+        return tree;
+    }
+
+    ast::ExpressionPtr postTransformSend(core::Context ctx, ast::ExpressionPtr tree) {
+        this->ignoreConstant = false;
+        return tree;
+    }
+
+    ast::ExpressionPtr postTransformConstantLit(core::Context ctx, ast::ExpressionPtr tree) {
+        if (this->ignoreConstant) {
+            return tree;
+        }
+
+        auto &lit = ast::cast_tree_nonnull<ast::ConstantLit>(tree);
+        if (!lit.symbol.isClassOrModule() && !lit.symbol.isFieldOrStaticField()) {
+            return tree;
+        }
+
+        auto loc = lit.symbol.loc(ctx);
+
+        auto otherFile = loc.file();
+        if (!otherFile.exists() || !otherFile.data(ctx).isPackaged()) {
+            return tree;
+        }
+
+        // If the imported symbol comes from the test namespace, we must also be in the test namespace.
+        if (otherFile.data(ctx).isPackagedTest() && !this->insideTestFile) {
+            if (auto e = ctx.beginError(lit.loc, core::errors::Packager::UsedTestOnlyName)) {
+                e.setHeader("Used test-only constant `{}` in non-test file", lit.symbol.show(ctx));
+            }
+        }
+
+        // no need to check visibility for these cases
+        auto otherPackage = ctx.state.packageDB().getPackageNameForFile(otherFile);
+        if (!otherPackage.exists() || this->package.mangledName() == otherPackage) {
+            return tree;
+        }
+
+        // Did we resolve a packaged symbol in unpackaged code?
+        if (!this->package.exists()) {
+            if (auto e = ctx.beginError(lit.loc, core::errors::Packager::PackagedSymbolInUnpackagedContext)) {
+                e.setHeader("Packaged constant `{}` used in an unpackaged context", lit.symbol.show(ctx));
+            }
+            return tree;
+        }
+
+        auto importType = this->package.importsPackage(otherPackage);
+        if (!importType.has_value()) {
+            // We failed to import the package that defines the symbol
+            if (auto e = ctx.beginError(lit.loc, core::errors::Packager::MissingImport)) {
+                auto &pkg = ctx.state.packageDB().getPackageInfo(otherPackage);
+                e.setHeader("No import provides `{}`", lit.symbol.show(ctx));
```

Comment:
```suggestion
                e.setHeader("`{}` exists but has not been imported", lit.symbol.show(ctx));
```

ditto the other comment; it looks like we chose to match the old error message, so i'm happy to push this to the future.